packages/gatsby/src/utils/webpack/cache-folder-resolver.ts

```diff
@@ -0,0 +1,65 @@
+import Resolver from "enhanced-resolve/lib/Resolver"
+
+interface IRequest {
+  request?: string
+  path: string
+}
+
+type ProcessWithPNP = NodeJS.ProcessVersions & { pnp?: string }
+
+/**
+ * To support PNP we have to make sure dependencies resolved from the .cache folder should be resolved from the gatsby package directory
+ */
+export class CacheFolderResolver {
+  private requestsFolder: string
+
+  constructor(requestsFolder: string) {
+    this.requestsFolder = requestsFolder
+  }
+
+  apply(resolver: Resolver): void {
+    if (!(process.versions as ProcessWithPNP).pnp) {
+      return
+    }
```

Comment:
Sorry if I'm missing some context here but shouldn't it always resolve it from `gatsby` since it owns the folder? Otherwise it's relying on hoisting quite a lot

Reply:
It should but it could be a subtle breaking change for some so I made sure to only fix it for pnp in v3